### PR TITLE
Auto-Close Answered Tickets

### PR DESCRIPTION
**Pull request by MarcBanyard**
This adds the ability to automatically close answered tickets after X amount of days of inactivity.

New Email Template created for "**Closed Tickets**"
The ability has also been added to the system to send Closed Ticket emails with the above template.

Both of the above can be set per department (**Admin Panel** > **Agents** > **Departments** > **[Department]**).

### CODE REVIEW

**Comment by ntozier** [comment]
There used to be two plugins that do this, I'm not sure that devs will accept this as a patch.
This might be functionality that they would want in a plugin not in core.

**Comment by MarcBanyard** [comment]
There have been several people on the forums that have asked for this over the years and it is such a valuable feature that we used, but when we upgraded from an older version it would not work, so we have re-written it.

Having it as a plugin is great, but plugins are quite often not maintained and the plugins we found did not work, so we have written it as part of the core code.

It has been written so that if you want it you can use it, if not it will not break any functionality of the system, so why would you not add it to the core especially since the plugins that were written no longer work?

**Comment by elekgeek** [comment]
Hope this feature gets added, thanks!

**Comment by JediKev** [comment]
Thank you for the pull request but we are currently refraining from merging any new functionality and making any major changes to the current codebase while we focus on v2.0 (full codebase rewrite).

**Comment by need4swede** [comment]
To confirm, the "days to close" counter only starts ticking after an initial reply from an agent, and resets after each reply? So it's only "x days of inactivity" after first reply? A bit confused about how that counter works. Thank you for your work!

**Comment by MarcBanyard** [comment]
> To confirm, the "days to close" counter only starts ticking after an initial reply from an agent, and resets after each reply? So it's only "x days of inactivity" after first reply? A bit confused about how that counter works. Thank you for your work!

Its a "Close after X days of inactivity once an agent has replied".
The counter only kicks in when an agent replies to a ticket, it is also reset each time an agent replies to the ticket.

So if you set it to 4 days and a customer emailed in, the following would be true.
- The ticket would stay open indefinitely as the counter only kicks in on an agent's reply.
- If an agent has replied to a ticket and the customer replies within 4 days it cancels the timer and the ticket remains open.
- If an agent has replied to a ticket and a customer does not reply within the 4 days the ticket is automatically closed.
- Once a ticket is closed a customer can email back in and the ticket is reopened, the above starts counting again.

Hope this makes sense.

**Comment by MarcBanyard** [comment]
> Thank you for the pull request but we are currently refraining from merging any new functionality and making any major changes to the current codebase while we focus on v2.0 (full codebase rewrite).

That's a pity.
I tried to ensure that there are no changes to the database and that not setting any of the options would not change any of the default behaviour of osTicket.

I know lots of people have asked for it on the forums previously and it would be great to get it merged into the core program as it does not cause any issues if it is not used.

This is the 2nd time I have re-written this into the code and the 2nd time I have requested it to be merged in.

I think osTicket is a great product and this feature would be of great benefit to be part of the core system as all of the paid for systems on the market have this as default.

**Comment by need4swede** [comment]
> > To confirm, the "days to close" counter only starts ticking after an initial reply from an agent, and resets after each reply? So it's only "x days of inactivity" after first reply? A bit confused about how that counter works. Thank you for your work!
> 
> Its a "Close after X days of inactivity once an agent has replied". The counter only kicks in when an agent replies to a ticket, it is also reset each time an agent replies to the ticket.
> 
> So if you set it to 4 days and a customer emailed in, the following would be true.
> 
> * The ticket would stay open indefinitely as the counter only kicks in on an agent's reply.
> * If an agent has replied to a ticket and the customer replies within 4 days it cancels the timer and the ticket remains open.
> * If an agent has replied to a ticket and a customer does not reply within the 4 days the ticket is automatically closed.
> * Once a ticket is closed a customer can email back in and the ticket is reopened, the above starts counting again.
> 
> Hope this makes sense.

Makes perfect sense. Thank you for the detailed response.

Curious to know if an additional feature could be added, where a ticket will auto close in X days when a ticket is set to a specific status, regardless of interaction. 

### **TLDR:**

Can this be done?

```
if (ticketStatus == 'Resolved'){ 
    onReplyTimerReset = false
    autoClose = true };

else { 
    onReplyTimerReset = true
    autoClose = true };
```

Let's say you have 'Open', 'Closed' and 'Resolved' states.

When an agent chooses 'Closed', it immediately closes the ticket, and the user can't reply / reopen the ticket. Pretty straightforward.

But the 'Resolved' state basically keeps the ticket open for X amount of days, so the user can reply to the thread, maybe a "Thank you" post of some kind - but it also activates a timer to close the ticket automatically. Once the timer is up, it will automatically move the ticket from 'Resolved' to 'Closed' and the user can longer reply to the thread and has to generate a new inquiry. 

This setting wouldn't have the timer reset automatically upon reply (only if ticket status == 'Resolved') and would require the agent to manually set the ticket to 'Open' for the timer to be disabled, but instead acts as a 'Delay Closing Ticket' method. In our case, 99% of the time the user will have a "Thank you" response within 24hrs of a ticket. I want them to have the ability to post that reply, but I don't want the ticket to move out of my 'Resolved' queue and I don't want the timer to reset after one of these replies. 

Would something like this be possible as an option for this? I don't have much experience with PHP and work mainly with JS, but I would think this can be incorporated.

**Comment by MarcBanyard** [comment]
> Curious to know if an additional feature could be added, where a ticket will auto close in X days when a ticket is set to a specific status, regardless of interaction.
> 
> ### **TLDR:**
> Can this be done?
> 
> ```
> if (ticketStatus == 'Resolved'){ 
>     onReplyTimerReset = false
>     autoClose = true };
> 
> else { 
>     onReplyTimerReset = true
>     autoClose = true };
> ```
> 
> Let's say you have 'Open', 'Closed' and 'Resolved' states.
> 
> But the 'Resolved' state basically keeps the ticket open for X amount of days, so the user can reply to the thread, maybe a "Thank you" post of some kind - but it also activates a timer to close the ticket automatically. Once the timer is up, it will automatically move the ticket from 'Resolved' to 'Closed' and the user can longer reply to the thread and has to generate a new inquiry.
> 
> This setting wouldn't have the timer reset automatically upon reply (only if ticket status == 'Resolved') and would require the agent to manually set the ticket to 'Open' for the timer to be disabled, but instead acts as a 'Delay Closing Ticket' method. In our case, 99% of the time the user will have a "Thank you" response within 24hrs of a ticket. I want them to have the ability to post that reply, but I don't want the ticket to move out of my 'Resolved' queue and I don't want the timer to reset after one of these replies.
> 
> Would something like this be possible as an option for this? I don't have much experience with PHP and work mainly with JS, but I would think this can be incorporated.

I Don't see why not, have a look at the files that were changed in the merge request and it should stand out, if you are able to read JS then you could have a tinker with PHP on a test site and see if that works for you.

You could duplicate the code from the files in my Merge Request and call it **Force Close Ticket After X Days**, there is a SQL query that will return the tickets that you could modify to suit your needs.


> When an agent chooses 'Closed', it immediately closes the ticket, and the user can't reply / reopen the ticket. Pretty straightforward.
> 
This is configurable by going to the **Admin Panel** > **Manage** > **Lists** > **Ticket Statuses** > **Closed** > **Item Properties** > Uncheck the **Allow tickets on this status to be reopened by end users** and click **Save**

**Comment by ntozier** [comment]
You can already do something like this.

Create a closed status (let's call it: *Closed-No-Reopen* at **Manage -> Lists -> Ticket Statuses**
Under the *Item Properties* tab. set State Closed, and uncheck "Allow tickets on this status to be reopened by end users"

Run an autoclose plugin, and any tickets taht you do not want re-opened set them to the *Closed-No-Reopen*.
When an Agent closes a ticket if they don't want it to be re-openable they would set the ticket to *Closed-No-Reopen*.

New replies to tickets that are not allowed to re-open generate a new ticket.

**Comment by need4swede** [comment]
> I Don't see why not, have a look at the files that were changed in the merge request and it should stand out, if you are able to read JS then you could have a tinker with PHP on a test site and see if that works for you.

Would something like this work?

"If ticket is set to 'Resolved' and has a status of 'open', automatically close ticket after X amount of days"

Screen Shot 2022-04-11 at 12 58 43

So now, regardless if an agent replies or not, as long as it's within the 'Resolved' queue, the ticket should auto-close after X days and still allow users to reply without resetting the timer.

EDIT: That didn't seem to work. Must've made an error somewhere.

**Comment by kurtzelda** [comment]
The AutoClose ticket don't work on 1.16.x (think is not php 8 compliant).
I think should be very very very ... (a lot of "very") fine having this function inside the original project.

**Comment by MarcBanyard** [comment]
The following Pull Request resolves the issue and works with PHP 8 and above
https://github.com/osTicket/osTicket/pull/6445

Hope this gets merged soon as it is extremely beneficial especially on busy systems by automatically closing tickets that customers have not replied to (i.e. the customers query has been resolved, so the ticket can be closed)

**Comment by philbertphotos** [comment]
I like this may manually added to my build.

**Comment by MarcBanyard** [comment]
> I like this may manually added to my build.

Please vote for this in #6445 as it would then be part of the core system.
Many people have asked for the ability over the years, so i've added it on several Pull Requests, but so far none have been merged in.

If many people like and need it there should be enough reason for the main devs to merge it into the core system.
Its been written to have no negateive impact to people who do not use it or do not want it, so merging it in would do no harm, but it would benefit everyone who does want and need it.

**Comment by kiongchiu82** [comment]
I need this feature, will stay tuned.

**Comment by asmith3006** [comment]
I would love to see this added!
